core/arch/arm/kernel/thread.c

```diff
@@ -1307,15 +1318,65 @@ static void thread_rpc_free(unsigned int bt, uint64_t cookie)
 	if (!get_rpc_arg(OPTEE_MSG_RPC_CMD_SHM_FREE, 1, &arg, &carg, &params))
 		return;
 
+	mobj_free(mobj);
+
 	params[0].attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
 	params[0].u.value.a = bt;
 	params[0].u.value.b = cookie;
 	params[0].u.value.c = 0;
-
 	reg_pair_from_64(carg, rpc_args + 1, rpc_args + 2);
 	thread_rpc(rpc_args);
 }
 
+static struct mobj *get_mobj_from_nested_params(struct optee_msg_param *params)
+{
+	paddr_t *pages = NULL;
+	paddr_t page_offset;
+	int num_pages;
+	struct mobj *nested_mobj = NULL;
+	struct optee_msg_param *nested_params;
+	struct mobj *mobj = NULL;
+
+	assert(params->attr == OPTEE_MSG_ATTR_TYPE_NESTED);
+
+	page_offset = params->u.nested.buf_ptr & SMALL_PAGE_MASK;
+
+	nested_mobj = msg_param_map_buffer(params->u.nested.buf_ptr,
+					   params->u.nested.params_num);
+	if (!nested_mobj)
+		goto out;
+
+	nested_params = mobj_get_va(nested_mobj, page_offset);
+	if (!msg_param_attr_is_tmem(nested_params)) {
+		mobj_free(nested_mobj);
```

Comment:
Remove, causes double free below `out:`
Fix the other `goto out` cases below too.